### PR TITLE
Cythonize udunits2

### PR DESCRIPTION
**Pull request by djkirkham**
Replaces uses of `ctypes` to access udunits with Cython code.

### CODE REVIEW

**Comment by marqh** [comment]
Hi @djkirkham 

interesting implementation, nice work

there are a collection of test failures on Python3 regarding strings and bytes

please have a look at
https://docs.python.org/2/c-api/string.html
https://docs.python.org/3/library/stdtypes.html

and clues such as 
http://stackoverflow.com/questions/7585435/best-way-to-convert-string-to-bytes-in-python-3
https://pythonhosted.org/six/

there should be a neat way of fixing most of these test issues

thank you

**Comment by djkirkham** [comment]
@marqh  Thanks, I resolved the issue by only allowing the wrapper library to accept byte strings. This is how it was working before anyway, so I haven't had to change the code in cf_units/__init__.py

**Comment by djkirkham** [comment]
The license header tests are beings skipped due to a bug. I've made a fix in https://github.com/SciTools/cf_units/pull/73, so it might be best to merge that PR first.

**Comment by marqh** [comment]
I think that this level of overhaul should go alongside a major version change; we should at least have considered it.
I have created a milestone for v2 and linked this PR to it

I'm working on a detail review, but I want to have decided about the version before this is merged.

So, any comments on whether this qualifies as a major version change are most welcome at this stage

thank you

**Comment by pelson** [comment]
This is awesome work @djkirkham. In principle I'm completely supportive of the change to move away from ctypes and embrace more traditional shared objects (especially as we observed such a dramatic impact of ctypes eventually calling gcc in order to find a shared library).

Agree with @marqh's sentiment - this should be a major version release, but that shouldn't be a big deal. The biggest blocker to merging this *right now* is having clear/good documentation on the migration path - there are public variables that have been removed (admittedly they look like an oversight, but they technically *are* public), and there are *new* public variables that have been added that also look like they probably shouldn't be... (```cf_units.FLOAT32``` anyone?).

In practice, we've been talking about how we improve our bandwidth on these kinds of issues, and ironically I think we are unlikely to be able to put a concerted effort on this for the next few months.

In short - I don't want to forget this, it is a really positive change! 🎉

File: cf_units/__init__.py [hunk 1]
```diff
@@ -1121,12 +880,6 @@ def __setstate__(self, state):
         #    handles aren't persistent)
         self.__init__(state['unit_text'], calendar=state['calendar'])
 
-    def __del__(self):
-        # NB. If Python is terminating then the module global "_ut_free"
-        # may have already been deleted ... so we check before using it.
-        if _ut_free:
-            _ut_free(self.ut_unit)
```
Comment (pelson) [inline cf_units/__init__.py:1128]: No longer needed because the cython wrapper does it for us 👍

File: cf_units/__init__.py [hunk 2]
```diff
@@ -1572,24 +1331,22 @@ def root(self, root):
             Taking a fractional root of a unit is not supported.
 
         """
-        try:
-            root = ctypes.c_int(root)
-        except TypeError:
-            raise TypeError('An int type for the root argument'
-                            ' is required')
-
+        if round(root) != root:
```
Comment (pelson) [inline cf_units/__init__.py:1334]: Functionally quite different. Think I'm OK with it though.

File: cf_units/__init__.py [hunk 3]
```diff
@@ -1572,24 +1331,22 @@ def root(self, root):
...
+            raise TypeError('An integer for the root argument is required')
         if self.is_unknown():
             result = self
         elif self.is_no_unit():
-            raise ValueError("Cannot take the logarithm of a 'no-unit'.")
+            raise ValueError("Cannot take the root of a 'no-unit'.")
```
Comment (pelson) [inline cf_units/__init__.py:1339]: Good catch.

File: cf_units/__init__.py [hunk 4]
```diff
@@ -1944,7 +1702,7 @@ def convert(self, value, other, ctype=FLOAT64, inplace=False):
             Value/s to be converted.
         * other (string/Unit):
             Target unit to convert to.
-        * ctype (ctypes.c_float/ctypes.c_double):
+        * ctype (cf_units.FLOAT32/cf_units.FLOAT64):
```
Comment (pelson) [inline cf_units/__init__.py:1705]: Tempting to get folks to use numpy here.

File: cf_units/__init__.py [hunk 5]
```diff
@@ -2002,49 +1759,45 @@ def convert(self, value, other, ctype=FLOAT64, inplace=False):
...
+                                         'convert to float or double.')
+                    # Utilise global convenience dictionary
+                    # _cv_convert_scalar
+                    result = _cv_convert_scalar[ctype](ut_converter,
+                                                       result)
+            return result
         else:
             raise ValueError("Unable to convert from '%r' to '%r'." %
                              (self, other))
-        return result
```
Comment (pelson) [inline cf_units/__init__.py:2047]: Let's return at the end.

File: cf_units/tests/test_unit.py [hunk 6]
```diff
@@ -290,14 +290,18 @@ def test_square_root(self):
         u = Unit('volt^2')
         self.assertEqual(u.root(2), 'V')
 
+    def test_square_root_integer_float(self):
+        u = Unit('volt^2')
+        self.assertEqual(u.root(2.), 'V')
```
Comment (pelson) [inline cf_units/tests/test_unit.py:295]: Added test feels good.

**Review by pelson — APPROVED** [review]
This change is epic, and awesome!

I'd like to move forwards with getting this merged and released as cf_units 2.0.

@SciTools/cf-units-devs: I'm happy to take responsibility for review+updates. How do you want to proceed in terms of managing the change. Do we want to do this slowly, or shall we just go for it?

**Comment by ocefpaf** [comment]
> Do we want to do this slowly, or shall we just go for it?

I'd say go for it :smile:

**Comment by bjlittle** [comment]
> Do we want to do this slowly, or shall we just go for it?

I'm keen that we bank this, so let's go for it. It has certainly been around for long enough!

We've just got the `travis-ci` tests working again, and I've nudged @kaedonkers so we'll bank #90 imminently... but did we want to tackle #85 before cutting `2.0` ?

**Comment by pelson** [comment]
Closed by #92.

**HUGE** appreciation for this @djkirkham. Was a really great change, that unfortunately stagnated.
We finally got it in, and will be looking to cut cf_units v2 at some point in the near future.
